File: algo/75. Sort Colors.js
// @9 Two Pointers

/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var sortColors = function(nums) {
    var swap = function(nums, index1, index2) {
        if (index1==index2) {
            return;
        }
        var temp = nums[index2];
        nums[index2] = nums[index1];
        nums[index1] = temp;
    }
    var left = 0;
    var right = nums.length-1;
    var cursor = 0;
    while (cursor <= right) {
        if (nums[cursor] == 1) {
            cursor++;
        } else if (nums[cursor] == 0) {
            swap(nums, left, cursor);
            cursor++;
            left++;
        } else {
            swap(nums, right, cursor);
            // do not increment cursor here since we don't know
            // what the new value is after swapping
            right--;
        }        
    }
};

File: round2/75. Sort Colors.cs
public class Solution {
    private void Swap(int[] nums, int index1, int index2) {
        int temp = nums[index1];
        nums[index1] = nums[index2];
        nums[index2] = temp;
    }
    public void SortColors(int[] nums) {
        int p1 = 0, p2 = nums.Length-1;
        for (int i=0; i<nums.Length; i++) {
            if (nums[i] == 0 && i!=p1) {
                Swap(nums, p1, i);
                p1++;
                i--;
            } else if (nums[i] == 2 && i<p2) {
                Swap(nums, p2, i);
                p2--;
                i--; 
            }
        }
    }
}